### PR TITLE
added waketags.py for creating tags file

### PR DESCRIPTION
**Pull request by sistillson**
Python script that leverages 'wake -g' to create a tags file for use with vi* and any other program that can read a tags file

### CODE REVIEW

**Comment by terpstra** [comment]
This is pretty cool!

File: share/doc/wake/syntax/tags/waketags.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+HEADER = \
+"""!_TAG_FILE_FORMAT       2       /extended format; --format=1 will not append ;" to lines/
+!_TAG_FILE_SORTED       1       /0=unsorted, 1=sorted, 2=foldcase/
+!_TAG_PROGRAM_AUTHOR    Chris Stillson  /stillson@sifive.com
+!_TAG_PROGRAM_NAME      waketags.py
+!_TAG_PROGRAM_VERSION   1.0     //"""
+
+FILE_RE_STR = r'(?P<sig>.*) = \<(?P<file>.*)\>$'
+FILE_RE = compile(FILE_RE_STR)
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: I tried combining it into a single regex, but the regex module seems to choke on unicode

File: share/doc/wake/syntax/tags/waketags.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+waketags.py
+"""
+
+
+HEADER = \
+"""!_TAG_FILE_FORMAT       2       /extended format; --format=1 will not append ;" to lines/
+!_TAG_FILE_SORTED       1       /0=unsorted, 1=sorted, 2=foldcase/
+!_TAG_PROGRAM_AUTHOR    Chris Stillson  /stillson@sifive.com
+!_TAG_PROGRAM_NAME      waketags.py
+!_TAG_PROGRAM_VERSION   1.0     //"""
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: sounds good, I will use the strip()

File: share/doc/wake/syntax/tags/waketags.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3.7
+import subprocess
+import sys
+from io import StringIO
+from pathlib import Path
+from re import compile
+
+"""
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: will fix

File: share/doc/wake/syntax/tags/waketags.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+Leverage 'wake -g' to great a tags file
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: will fix

File: share/doc/wake/syntax/tags/waketags.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+The main weakness of this approach is that it uses line numbers
+instead of regular expressions to find tags. Meaning frequent re-runs
+to keep it up to date. Also, this only finds globals.
+
+usage:
+make sure that wake-db exists
+in the top of your workspace:
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: comments need work, agreed. will clean up

File: share/doc/wake/syntax/tags/waketags.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+if __name__ == '__main__':
+
+    output = StringIO()
+
+    for a_line in get_wake_globals():
+        try:
+            tag, rest = a_line.split(':', 1)
+        except ValueError:
+            print('badline: %s' % a_line, sys.stderr)
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: yes, good catch

File: share/doc/wake/syntax/tags/waketags.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            continue
+
+        try:
+            descr, newtag = tag.split(" ", 1)
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: argreed. missed that during dev.

File: share/doc/wake/syntax/tags/waketags.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
+#!/usr/bin/env python3.7
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: will change

File: share/doc/wake/syntax/tags/waketags.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+            tag = newtag
+
+        spec = FILE_RE.match(rest)
+        file_spec = spec.group('file').strip()
+        signature = spec.group('sig').strip()
+        file_name, file_line, location = file_spec.split(':')
+        if file_line.startswith('['):
+            file_line = file_line.split('-')[0][1:]
+
+        print('%s\t%s\t%s;"' % (tag, file_name, file_line), file=output)
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: I originally just printed straight to the tags file, got a little too focused in

File: share/doc/wake/syntax/tags/waketags.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+    tagfile = Path('tags').open(mode='w')
+
+    print(HEADER, file=tagfile)
+
+    output.seek(0)
+    for out_line in sorted(output.readlines()):
```
[the remark this answers is not in the document]
Comment (sistillson) [reply share/doc/wake/syntax/tags/waketags.py]: that's the right answer. will fix

**Comment by davidmlw** [comment]
Hi, I just file a PR here https://github.com/sifive/wake/pull/384.
wake --ctags and wake --etags can generate tag files for vim and emacs. I just test etags with emacs.

**Comment by terpstra** [comment]
How does this relate to #445?

**Comment by terpstra** [comment]
Closing in favor of #445

**Pull request closed without merging**